cpp/include/raft/label/classlabels.cuh

```diff
@@ -0,0 +1,193 @@
+ * \param [in] n number of labels
+ * \param [in] y_unique device array of unique labels, size [n_classes]
+ * \param [in] n_classes number of unique labels
+ * \param [out] y_out device array of output labels
+ * \param [in] idx index of unique label that should be labeled as 1
+ * \param [in] stream cuda stream
+ */
+template <typename math_t>
+void getOvrlabels(math_t *y, int n, math_t *y_unique, int n_classes,
+                  math_t *y_out, int idx, cudaStream_t stream) {
+  ASSERT(idx < n_classes,
+         "Parameter idx should not be larger than the number "
+         "of classes");
+  raft::linalg::unaryOp(
+    y_out, y, n,
+    [idx, y_unique] __device__(math_t y) {
+      return y == y_unique[idx] ? +1 : -1;
+    },
+    stream);
+  CUDA_CHECK(cudaPeekAtLastError());
+}
+
+// TODO: add one-versus-one selection: select two classes, relabel them to
+// +/-1, return array with the new class labels and corresponding indices.
+
+template <typename Type, int TPB_X, typename Lambda>
+__global__ void map_label_kernel(Type *map_ids, size_t N_labels, Type *in,
+                                 Type *out, size_t N, Lambda filter_op,
+                                 bool zero_based = false) {
+  int tid = threadIdx.x + blockIdx.x * TPB_X;
+  if (tid < N) {
+    if (!filter_op(in[tid])) {
+      for (size_t i = 0; i < N_labels; i++) {
+        if (in[tid] == map_ids[i]) {
+          out[tid] = i + !zero_based;
+          break;
+        }
+      }
+    }
+  }
+}
+
+/**
+   * Maps an input array containing a series of numbers into a new array
+   * where numbers have been mapped to a monotonically increasing set
+   * of labels. This can be useful in machine learning algorithms, for instance,
+   * where a given set of labels is not taken from a monotonically increasing
+   * set. This can happen if they are filtered or if only a subset of the
+   * total labels are used in a dataset. This is also useful in graph algorithms
+   * where a set of vertices need to be labeled in a monotonically increasing
+   * order.
+   * @tparam Type the numeric type of the input and output arrays
+   * @tparam Lambda the type of an optional filter function, which determines
+   * which items in the array to map.
+   * @param out the output monotonic array
+   * @param in input label array
+   * @param N number of elements in the input array
+   * @param stream cuda stream to use
+   * @param filter_op an optional function for specifying which values
+   * should have monotonically increasing labels applied to them.
+   */
+template <typename Type, typename Lambda>
+void make_monotonic(Type *out, Type *in, size_t N, cudaStream_t stream,
+                    Lambda filter_op,
+                    std::shared_ptr<raft::mr::device::allocator> allocator,
+                    bool zero_based = false) {
+  static const size_t TPB_X = 256;
+
+  dim3 blocks(raft::ceildiv(N, TPB_X));
+  dim3 threads(TPB_X);
+
+  Type *map_ids;
```

Comment:
This should be a 0 sized device buffer that is passed by reference and resized later

Reply:
I'm not sure if I completely agree w/ this one but I'd like to minimize deltas here as much as possible since we're making the move before changing the code in cuml.